schemas: clean schemas

jacquerie opened the pull request:
Continues #1069, #1070, and #1143. Here's the plan:


- [x] Fix imports from `inspirehep.dojson.utils`
- [x] Implement `remove_duplicates` from `inspirehep.dojson.utils` in terms of `inspirehep.utils.dedupers`
- [x] Implement `remove_duplicates` from `inspirehep.modules.theme` in terms of `inspirehep.utils.dedupers`
- [x] Move `inspirehep/dojson/utils.py` in `inspirehep/dojson/utils/__init__.py`
- [x] Add @tomaszgy's non-geo helpers to `inspirehep/dojson/utils/__init__.py`, with tests
- [x] Add @bittirousku's helpers to `inspirehep/dojson/utils/__init__.py`, with tests
- [x] Add `inspirehep/dojson/utils/geo.py` from @tomaszgy, with tests
- [x] Add element schemas in `inspirehep/modules/records/jsonschemas/records/elements`, add test to check their validity
- [x] Amend schemas in `inspirehep/modules/records/jsonschemas` to use elements, collection by collection (with DoJSON rules and tests)

    - [x] Address
    - [x] Contact
    - [x] Fields
    - [x] ISBN & ISSN
    - [x] Ranks
    - [x] Titles
    - [x] URLs

- [x] Final cleanup

Extra:
- Fix DoJSON docstrings
- Test current methods in `inspirehep/dojson/utils.py`
- See if `inspirehep/dojson/utils/geo.py` can be better implemented in terms of `pycountry`
- Refactor DoJSON tests
- Integrate #1070

jacquerie commented:
TODO: ping #1178 when this is merged.

jacquerie commented on inspirehep/dojson/hep/fields/bd70x75x.py:
```diff
@@ -24,9 +24,10 @@
 
 from dojson import utils
```
TODO: `__future__` block.

jacquerie commented on inspirehep/dojson/hep/fields/bd76x78x.py:
```diff
@@ -24,9 +24,8 @@
 
 from dojson import utils
```
TODO: `__future__` block and license year.

jacquerie commented on inspirehep/modules/records/jsonschemas/records/hep.json:
```diff
@@ -518,35 +518,8 @@
-            },
-            "type": "array"
+        "field_categories": {
+            "$ref": "elements/fields.json"
```
I think it's better if elements define singular stuff (a field, instead of fields) so that they are more flexible. One can always write
```JSON
{
    "field_categories": {
        "type": "array",
        "items": {"$ref": "elements/field.json"},
        "uniqueItems": true
    }
}
```
Since this is how it was originally, what was the reason to change it to plural? @tomaszgy @bittirousku @kaplun

(Same thing applies to titles)

kaplun replied:
IMHO, in general, for certain type of categories such as the above `fields`, to allow for singular form is only going to cause troubles in the future when we realize that nevertheless the field is always repeatable.

jacquerie commented:
Travis failure might be unrelated: https://travis-ci.org/inspirehep/inspire-next/jobs/134946278#L2042 @jalavik

jacquerie commented:
Failure in https://github.com/inspirehep/inspire-next/pull/1170#issuecomment-223502771 was **not** unrelated: one step of the workflow was checking for the presence of stuff in `subject_terms`, which got renamed to `field_categories`.

So expect a lot of breakage after this PR is merged, because this was really just the tip of the iceberg. I'll try to amend all uses I can find.

kaplun commented:
Following @annetteholtkamp reports in [Asana](https://app.asana.com/0/2345936208876/115950233761060/f), we should ensure that:
* [ ] Experiment field in HEP is using the `yield` decorator
* [ ] Coden (084__a) are using the `yield` decorator
* [ ] 037__a as well
* [ ] 980__a as well

jacquerie commented:
The `yield` decorator is just one way of achieving that thing: one can also use side effects on the `self` that is passed.

BTW, can we solve those in another issue? Here I'm already having a hard time integrating what was already done...

kaplun commented:
> BTW, can we solve those in another issue? Here I'm already having a hard time integrating what was already done...

As you prefer. Was just putting it here, in case you were already amending these very rules.

jacquerie commented:
> As you prefer. Was just putting it here, in case you were already amending these very rules.

The schemas definitely need some love, and to go from DoJSON to DoneJSON.

jacquerie commented:
# ! Bumpy Road Ahead ! #

This PR changes schemas and enough uses to make the tests pass, but definitely not all the uses. Here's a compendium of what changed, with the symbol **+** denoting that the field was _added_, **-** denoting that the field was _deleted_, **~** denoting that the internal structure _changed_.

## Authors ##

* ~ `field_categories`: added `_term`, `term`, `_scheme`, `scheme`, and `source`, removed `name` subfields

## Conferences ##

* + `address`
* + `contact_details`
* - `contact_email`
* - `contact_person`
* - `field_code`
* + `field_categories`
* ~ `alternative_titles`: added `source`, `title`, and `subtitle` subfields
* - `subtitle`
* - `title`
* + `titles`
* - `proceedings`
* - `transparencies`
* - `url`
* + `urls`

## Experiments ##

* - `contact_details`
* - `contact_email`
* - `name`
* + `experiment_names`
* - `other_names`
* + `other_experiment_names`
* - `title`
* + `titles`
* - `url`
* + `urls`

## HEP ##

* ~ `authors`: removed `profile` subfield
* - `subject_terms`
* + `field_categories`
* ~ `title_variations`: added `source`, `title`, and `subtitle` subfields
* ~ `title_translations`: added `source` subfield
* ~ `urls`: removed `url`, `doc_string`, `material_type`, `size`, `comment`, and `name` subfields

## Institutions ##

* ~ `address`: added `state`, `latitude`, and `longitude`, removed `country` and `state_province`
* - `content_classification`
* + `field_categories`
* ~ `urls`: added `value` and `description` subfields

## Jobs ##

* + `contact_details`
* - `contact_email`
* - `contact_person`
* - `research_area`
* + `field_categories`
* ~ `urls`: added `value` and `description` subfields

## Journals ##

* - `former_title`
* - `inspire_long_title`
* - `inspire_short_title`
* + `short_titles`
* - `journal_title`
* + `journal_titles`
* ~ `title_variants`: added `source`, `title`, and `subtitle` subfields
* - `license_url`
* + `license_urls`
* - `url`
* + `urls`

jacenkow commented:
@jacquerie well done 🔥
